src/self_encryptor.rs

```diff
@@ -227,65 +204,184 @@ where
 
     /// Truncate the self_encryptor to the specified size (if extended, filled with `0u8`s).
     pub async fn truncate(&self, new_size: u64) -> Result<(), SelfEncryptionError> {
```

Comment:
this function now becomes bit hard to follow the process flow.
I'd suggest have :
1, have some helper function to reduce the duplicated blocks
2, add a detailed comment (better with a table) to explain what happens under different branches

Reply:
Yes I anticipated this problem. I will do this change